src/modules/ModuleRunner.ts

```diff
@@ -66,13 +82,67 @@ export class ModuleRunner {
         return merged;
     }
 
+    /**
+     * Ensure we register extensions provided by the modules
+     */
+    private updateExtensions(): void {
+        const cryptoSetupExtensions: Array<RuntimeModule> = [];
+        const experimentalExtensions: Array<RuntimeModule> = [];
+
+        this.modules.forEach((m) => {
+            /* Record the cryptoSetup extensions if any */
+            if (m.module.extensions?.cryptoSetup) {
+                cryptoSetupExtensions.push(m.module);
+            }
+
+            /* Record the experimantal extensions if any */
+            if (m.module.extensions?.experimental) {
+                experimentalExtensions.push(m.module);
+            }
+        });
+
+        /* Enforce rule that only a single module may provide a given extension */
+        if (cryptoSetupExtensions.length > 1) {
+            throw new Error(
+                `cryptoSetup extension is provided by modules ${cryptoSetupExtensions
+                    .map((m) => m.moduleName)
+                    .join(", ")}, but can only be provided by a single module`,
+            );
+        }
+        if (experimentalExtensions.length > 1) {
+            throw new Error(
+                `experimental extension is provided by modules ${experimentalExtensions
+                    .map((m) => m.moduleName)
+                    .join(", ")}, but can only be provided by a single module`,
+            );
+        }
+
+        /* Override the default extension if extension was provided by a module */
+        if (cryptoSetupExtensions.length == 1) {
+            this.extensions.cryptoSetup = cryptoSetupExtensions[0].extensions?.cryptoSetup;
+        }
+
+        if (experimentalExtensions.length == 1) {
+            this.extensions.experimental = cryptoSetupExtensions[0].extensions?.experimental;
+        }
+    }
+
     /**
      * Registers a factory which creates a module for later loading. The factory
      * will be called immediately.
```

Comment:
Existing issue, but I don't really understand this.

> creates a module for **later** loading

vs

>  will be called **immediately**.

These two seem to be contradictory? Is the module loaded later, or immediately? Or is "loading" some separate step that happens after the module is created? I can't see it, if so?

Reply:
The code and the comment was written by @turt2live a while back.  It is a bit convoluted, and I don't know the exact rationale for doing it this way (but I'm sure there is one, probably to do with how modules are registered by element-web, and not matrix-react-sdk)

But the code says that there is no late-loading since: 
* Calling registerModule immediately creates a new AppModule
* The AppModule immediately calls the factory (in its constructor)